### DISCUSSION TITLE
Handling UnicodeErrors

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#11605] [Using MicroPython] [answered]
I am getting unicode errors on a LoRa link and can't trap these errors.  I am aware that there is not a specific UnicodeError in Micropython, but thought OSError caught all exceptions.

This doesn't seem to work:
```py
    try:
        PV_msg = msg.decode('utf-8')
        print(f'the message is {PV_msg}')
    except OSError as error:
        try:
            with open('errors.txt', 'a') as outfile:
                outfile.write('bad message after decode: ' + str(error) + '\n')
        except OSError:
            pass
```

**Comment by jimmo** [comment]
> but thought OSError caught all exceptions

This is not the case. It catches exceptions that come from OS-level functionality (e.g. filesystem, process, etc).

The line `except OSError:` catches OSError and anything derived from OSError.

If you want to catch all normal exceptions, then you want `except Exception:`, or if you want to catch absolutely anything then just plain
`except:` (without a type) is an option.

Although in this case I suspect you want to use `except UnicodeError:` ?

See
https://docs.python.org/3/library/exceptions.html

But also you might find it easier to see the hierarchy here: https://github.com/micropython/micropython/blob/master/py/objexcept.c#L304

**Comment by jimmo** [comment, marked as answer]
As a follow-up of sorts to the previous thread, I do not recommend this as a way of detecting transmission errors as you can still get corrupted/malformed messages that will pass this check. You should have some sort of proper message integrity check, e.g. a checksum.

**Reply by davefes** [reply]
Thank you.  UnicodeError wasn't in this doc:
https://docs.micropython.org/en/latest/library/builtins.html
I will take your advice on checksum and try enabling CRC.
